### ISSUE TITLE
[DependencyInjection] Allow registering attributes on classes that cannot be instantiated

### ISSUE THREAD

**Issue by ruudk** [symfony/symfony#48392]
### Description

In our application we rely Symfony's `ContainerBuilder::registerAttributeForAutoconfiguration` together with auto wire for a lot of things. I also helped improving it:
- https://github.com/symfony/symfony/pull/42039

But now I have an issue. I have a class that has a `private` constructor. It needs to be constructed by calling a static method. 
Example:
```php
#[SomeAttribute]
final class SomeService
{
    private function __construct() {}
}
```

I have a compiler pass that tries to work this class by looking for `SomeAttribute`:
```php
$container->registerAttributeForAutoconfiguration(
    SomeAttribute::class,
    static function (ChildDefinition $definition, SomeAttribute $attribute, ReflectionClass $reflector) : void {
        // do magic
    },
);
```

I load everything using:
```
$services->load('App\\', __DIR__ . '/');
```

Because of this line the class is never loaded because it's not instantiable.
https://github.com/symfony/symfony/blob/69f46f231b16be1bce1e2ecfa7f9a1fb192cda22/src/Symfony/Component/DependencyInjection/Loader/FileLoader.php#L232-L234

Possible solutions:

1. Always run the registered attributes configurators for classes that cannot be instantiated
2. Allow an additional option on the `PrototypeConfigurator` to indicate you want to always load attributes
3. Change the condition above:
    ```diff
    - if ($r->isInstantiable() || $r->isInterface()) {
    -    $classes[$class] = null;
    - }
    + $classes[$class] = null;
    ```

/cc @nicolas-grekas since we worked in this area together in #42039 do you have ideas?

**Comment by nicolas-grekas** [comment]
Shouldn't you use `#[Autoconfigure(factory: etc)]` on your class instead of using a compiler pass? That's not possible currently, but maybe we can improve this?

**Comment by ruudk** [comment]
@nicolas-grekas Yeah this is just an example. The point I was making is that the behaviour broke but you seem to have fixed it already https://github.com/symfony/symfony/issues/48411 🎉 
I'm pretty sure this will solve it and will test it out when 6.2.1 is released. 
Thanks!

**Comment by ruudk** [comment]
Turns out I still encounter this issue. Now I have a better example + possible solution.

We use a lot of Identity objects. Here is an example of an `OrderId` object. 

```php
#[SomeAttribute]
final class OrderId implements Stringable
{
    // This constructor is private, we want to force the use of the named constructors (create, fromString).
    private function __construct(
        private UuidInterface $id,
    ) {}

    public static function create() : self
    {
        return new self(Uuid::uuid4());
    }

    public static function fromString(string $id) : self
    {
        return new self(Uuid::fromString($id));
    }

    public function __toString() : string
    {
        return (string) $this->id;
    }
}

// Somewhere in a bundle:
$container->registerAttributeForAutoconfiguration(
    SomeAttribute::class,
    static function (ChildDefinition $definition, SomeAttribute $attribute, ReflectionClass $reflector) : void {
        // do something with the class, for example, gather them in an array and attach them to another service.
    },
);
```

As you can see, the constructor of this is private. For some developers, this makes sense, because they want people to use the named constructors only. The `OrderId` is not intended to be a service, ever. 

But we want to apply some magic to them. We want to annotate them with `#[SomeAttribute]`, and then use `registerAttributeForAutoconfiguration` to find them all. 

But this is not possible. Classes that are not instantiable are skipped. But it's very hard for a developer to find out _why_ this is not working. 

## Solution

This is done here:
https://github.com/symfony/symfony/blob/6c446841d5251e03226ac7fe8b8d0293dd9a6119/src/Symfony/Component/DependencyInjection/Loader/FileLoader.php#L295-L297

Turns out, there is one exception to this rule, and that's the `Autoconfigure` detection:
https://github.com/symfony/symfony/blob/6c446841d5251e03226ac7fe8b8d0293dd9a6119/src/Symfony/Component/DependencyInjection/Loader/FileLoader.php#L299-L301

Can we alter `registerAttributeForAutoconfiguration`, so that a developer can say "give me also classes that are not instantiable"?

```php
$container->registerAttributeForAutoconfiguration(
    SomeAttribute::class,
    static function (ChildDefinition $definition, SomeAttribute $attribute, ReflectionClass $reflector, findAll: true) : void {
        // `$findAll = true`, this will not present everything: classes with private constructor, even interfaces can show up.
    },
);
```

**Comment by nicolas-grekas** [comment]
Can we skip adding any options and instead tag such classes as container.excluded? Would that help?

**Comment by ruudk** [comment]
Smart. I played a bit with it and created a proof of concept PR: 
- https://github.com/symfony/symfony/pull/53605

This works for my use case. But I wonder if this is what you meant.
